### PR TITLE
Use compatible Python executable installed by Pyenv

### PR DESCRIPTION
**Pull request by ggicci**
# Pull Request Check List

Resolves:  #4079 #4101

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

A new config item `virtualenvs.use-pyenv` with default value `true` has been instroduced. 

### Integrate `pyenv` with `poetry`'s discovery logic

If no compatible versions were found in the system, `poetry` will keep looking for compatible Python interpreters installed by `pyenv`.

## Fixes

1. #4101 Create virtualenvs by passing the absolute path to the Python executable to `virtualenv`.
2. Incorrect sorting algorithm to sort the available Python versions.

### CODE REVIEW

**Comment by ggicci** [comment]
Test output here:

```bash
$ python2 -V
Python 2.7.17

$ python3 -V
Python 3.6.9

$ pyenv versions
* system (set by /home/mingjietang/.pyenv/version)
  3.7.10
  3.8.10
  3.9.4

$ cd ~/workspace/src/ggicci.me/poetry-pyenv-test
$ grep "python =" pyproject.toml
python = "~3.7"

$ poetry@local -vvv install
Loading configuration file /home/ggicci/.config/pypoetry/config.toml
The currently activated Python version 3.6.9 is not supported by the project (~3.7).
Trying to find and use a compatible version. 
Pyenv loaded: /home/ggicci/.pyenv/bin/pyenv
Trying python3
Trying python3.7
Trying /home/ggicci/.pyenv/versions/3.7.10/bin/python3.7
Using /home/ggicci/.pyenv/versions/3.7.10/bin/python3.7 (3.7.10)
Creating virtualenv poetry-pyenv-test-gx4qtdVT-py3.7 in /home/ggicci/.cache/pypoetry/virtualenvs
Using virtualenv: /home/ggicci/.cache/pypoetry/virtualenvs/poetry-pyenv-test-gx4qtdVT-py3.7
Installing dependencies from lock file

Finding the necessary packages for the current system
```

**Review by Leonardofreua — CHANGES_REQUESTED** [review]

File: poetry/utils/env.py [hunk 1]
```diff
@@ -925,6 +868,56 @@ def create_venv(
...
+
+    def _list_sys_python_candidates(self) -> List[str]:
+        candidates = []
+        supported_python = self._poetry.package.python_constraint
+        for python_to_try in sorted_trying_versions(
+            self._poetry.package.AVAILABLE_PYTHONS  # "3", "3.5", "3.6", "2", "2.7", ...
+        ):
+            if "." not in python_to_try:
+                if parse_constraint(f"^{python_to_try}.0").allows_any(supported_python):
+                    candidates.append("python" + python_to_try)
```
[the remark this answers is not in the document]
Comment (ggicci) [reply poetry/utils/env.py:905]: Thanks for this good suggestion. I never noticed this tip of performance improvement before.

File: poetry/utils/pyenv.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+class Pyenv:
+    def __init__(self) -> None:
+        self._command = None
+        self._versions = None
+
+    def __bool__(self) -> bool:
+        return self._command is not None
+
+    def load(self) -> None:
+        if self._command is not None:
```
[the remark this answers is not in the document]
Comment (ggicci) [reply poetry/utils/pyenv.py:22]: I prefer putting "fast fail" branches of statements in the head of a method and leaving "main logic" at the last.

**Comment by NickKaramoff** [comment]
@ggicci why did you close this PR?

**Comment by ggicci** [comment]
This pr brought too many changes and should break down into small pieces.

**Pull request closed without merging**